cloud/amazon/ec2_customer_gateway.py

```diff
@@ -0,0 +1,239 @@
+        self.module = module
+
+        try:
+            region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module, boto3=True)
+            if not region:
+                module.fail_json(msg="Region must be specified as a parameter, in EC2_REGION or AWS_REGION environment variables or in boto configuration file")
+            self.ec2 = boto3_conn(module, conn_type='client', resource='ec2', region=region, endpoint=ec2_url, **aws_connect_kwargs)
+        except boto.exception.NoAuthHandlerFound, e:
+            module.fail_json(msg="Can't authorize connection - " + str(e))
+
+    def ensure_cgw_absent(self, gw_id):
+        response = self.ec2.delete_customer_gateway(
+            DryRun=False,
+            CustomerGatewayId=gw_id
+        )
+        return response
+            
+    def ensure_cgw_present(self, bgp_asn, ip_address):
+        response = self.ec2.create_customer_gateway(
+            DryRun=False,
+            Type='ipsec.1',
+            PublicIp=ip_address,
+            BgpAsn=bgp_asn,
+        )
+        return response
+
+    def tag_cgw_name(self, gw_id, name):
+        response = self.ec2.create_tags(
+            DryRun=False,
+            Resources=[
+                gw_id,
+            ],
+            Tags=[
+                {
+                    'Key': 'Name',
+                    'Value': name
+                },
+            ]
+        )
+        return response
+
+    def describe_gateways(self, ip_address):
+        response = self.ec2.describe_customer_gateways(
+            DryRun=False,
+            Filters=[
+                {
+                    'Name': 'state', 
+                    'Values': [
+                        'available',
+                    ]
+                },
+                {
+                    'Name': 'ip-address', 
+                    'Values': [
+                        ip_address,
+                    ]
+                }
+            ]
+        )
+        return response
+
+def main():
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(
+        dict(
+            bgp_asn = dict(required=False, type='int'),
+            ip_address = dict(required=True),
+            name = dict(required=True),
+            state = dict(default='present', choices=['present', 'absent']),
+        )
+    )
+
+    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=True)
+
+    if not HAS_BOTOCORE:
+        module.fail_json(msg='botocore is required.')
+
+    if not HAS_BOTO3:
+        module.fail_json(msg='boto3 is required.')
+
+    gw_mgr = Ec2CustomerGatewayManager(module)
+
+    bgp_asn = module.params.get('bgp_asn')
+    ip_address = module.params.get('ip_address')
+    name = module.params.get('name')
+
+    existing = gw_mgr.describe_gateways(module.params['ip_address'])
+
+    results = dict(changed=False)
+    if module.params['state'] == 'present':
+        if existing['CustomerGateways']:
+            results['gateway']=existing
```

Comment:
Specifically, I need to evaluate the contents of existing, find the tag named "Name" if it exists, and if it doesn't exist or it's value is not equal to the name variable then I need to call the tag function.  Doing that evaluation is a bit beyond my current python skills.  Below is an example of the data returned.  CustomerGateway will only ever contain one item if the desired ip address already exists.

```
{
    "changed": false, 
    "gateway": {
        "CustomerGateways": [
            {
                "BgpAsn": "65000", 
                "CustomerGatewayId": "cgw-9a59b6f3", 
                "IpAddress": "1.2.3.4", 
                "State": "available", 
                "Tags": [
                    {
                        "Key": "Name", 
                        "Value": "foobar"
                    }
                ], 
                "Type": "ipsec.1"
            }
        ], 
        "ResponseMetadata": {
            "HTTPStatusCode": 200, 
            "RequestId": "7a3f781f-1c68-4169-a709-ee41ea4e8d5f"
        }
    }, 
    "invocation": {
        "module_args": {
            "aws_access_key": null, 
            "aws_secret_key": null, 
            "bgp_asn": 65000, 
            "ec2_url": null, 
            "ip_address": "1.2.3.4", 
            "name": "foobar1", 
            "profile": null, 
            "region": "us-east-1", 
            "security_token": null, 
            "state": "present", 
            "validate_certs": true
        }
    }
}
```